File: leetcode/684-redundant-connection/main.py
class Solution(object):
    def findRedundantConnection(self, edges):
        """
        :type edges: List[List[int]]
        :rtype: List[int]

        1st approach: Union Find
        - implement an Union Find
        - for each edge, union the node if they are not connected
        - but if they both present in the union set, put it into the result

        The purpose of this question is to practice Union Find, I am gonna do it in other ways

        Time    O(nlogn) iterate the edges, in each iteration the Find() operation is O(logn)
        Space   O(n) the set
        28 ms, faster than 69.39% 
        """
        nodesSet = set()
        for edge in edges:
            nodesSet.add(edge[0])
            nodesSet.add(edge[1])
        nodes = list(nodesSet)
        uf = UnionFind(nodes)
        res = None
        for edge in edges:
            temp = uf.union(edge[0], edge[1])
            if temp == False:
                res = edge
        if uf.getCount() > 1:
            return []
        return res


class UnionFind(object):
    def __init__(self, vertices):
        self.count = len(vertices)
        self.ids = {}
        self.caps = {}
        for vertex in vertices:
            self.ids[vertex] = vertex
            self.caps[vertex] = 1

    def getCount(self):
        return self.count

    # loop to find to ultimate root
    def find(self, key):
        cur = key
        while cur != self.ids[cur]:
            cur = self.ids[cur]
        return cur

    # attach to a larger tree
    def union(self, p, q):
        pId = self.find(p)
        qId = self.find(q)
        if pId == qId:
            return False

        if self.caps[pId] < self.caps[qId]:
            self.ids[pId] = qId
            self.caps[qId] += self.caps[pId]
        else:
            self.ids[qId] = pId
            self.caps[pId] += self.caps[qId]
        self.count -= 1
        return True

    def isConnect(self, p, q):
        return self.find(p) == self.find(q)


print(Solution().findRedundantConnection([[1, 2], [1, 3], [2, 3]]))
print(Solution().findRedundantConnection(
    [[1, 2], [2, 3], [3, 4], [1, 4], [1, 5]]))
print(Solution().findRedundantConnection(
    [[1, 2], [2, 3], [3, 4], [1, 4], [1, 5], [4, 5]]))
print(Solution().findRedundantConnection(
    [[1, 2], [2, 3], [3, 4], [1, 4], [1, 5], [6, 7]]))

File: leetcode/684-redundant-connection/main.go
package main

import "fmt"

/*
	1st approach: Union Find
	- implement an Union Find
	- for each edge, union the node if they are not connected
	- but if they both present in the union set, put it into the result

	The purpose of this question is to practice Union Find, I am gonna do it in other ways

	Time    O(nlogn) iterate the edges, in each iteration the Find() operation is O(logn)
	Space   O(n) the set
	8 ms, faster than 53.57%
*/
func findRedundantConnection(edges [][]int) []int {
	nodesSet := make(map[int]bool)
	for _, edge := range edges {
		nodesSet[edge[0]] = true
		nodesSet[edge[1]] = true
	}
	nodes := []int{}
	for key := range nodesSet {
		nodes = append(nodes, key)
	}
	uf := Constructor(nodes)
	res := []int{}
	for _, edge := range edges {
		temp := uf.Union(edge[0], edge[1])
		if temp == false {
			res = edge
		}
	}
	if uf.GetCount() > 1 {
		return []int{}
	}
	return res
}

type UnionFind struct {
	Count int
	Ids   map[int]int
	Caps  map[int]int
}

func Constructor(nodes []int) UnionFind {
	ids := make(map[int]int)
	caps := make(map[int]int)
	for _, node := range nodes {
		ids[node] = node
		caps[node] = 1
	}
	return UnionFind{len(nodes), ids, caps}
}

func (this *UnionFind) GetCount() int {
	return this.Count
}

// O(logN): loop to find to ultimate root
func (this *UnionFind) Find(key int) int {
	cur := key
	for cur != this.Ids[cur] {
		cur = this.Ids[cur]
	}
	return cur
}

// union to the bigger tree, O(N)
func (this *UnionFind) Union(p int, q int) bool {
	pId := this.Find(p)
	qId := this.Find(q)
	if pId == qId {
		return false
	}
	if this.Caps[pId] < this.Caps[qId] {
		this.Ids[pId] = qId
		this.Caps[qId] += this.Caps[pId]
	} else {
		this.Ids[qId] = pId
		this.Caps[pId] += this.Caps[qId]
	}
	this.Count--
	return true
}

func (this *UnionFind) IsConnect(p int, q int) bool {
	return this.Find(p) == this.Find(q)
}

func main() {
	a := [][]int{
		{1, 2}, {1, 3}, {2, 3},
	}
	fmt.Println(findRedundantConnection(a))

	a = [][]int{
		{1, 2}, {2, 3}, {3, 4}, {1, 4}, {1, 5},
	}
	fmt.Println(findRedundantConnection(a))

	a = [][]int{
		{1, 2}, {2, 3}, {3, 4}, {1, 4}, {1, 5}, {4, 5},
	}
	fmt.Println(findRedundantConnection(a))

	a = [][]int{
		{1, 2}, {2, 3}, {3, 4}, {1, 4}, {1, 5}, {6, 7},
	}
	fmt.Println(findRedundantConnection(a))
}
